### ISSUE TITLE
Parent job does not execute when sharing dependency with another job

### ISSUE THREAD

**Issue by bilalshaikh42** [taskforcesh/bullmq#1021]
We have a series of processing steps linked to particular Id. Some of these steps depends on others, and some dependencies are shared. We have defined a flow as follows: 

```typescript

const runId =SomeSharedID
    
const JobA = {
      name: 'A',
      queueName: QueueA
      opts: {
        jobId: `A-${runId}`,
      },
      data: {
        runId: runId,
      },
    };

    const JobB = {
      name: 'B',
      queueName: QueueB
      opts: {
        jobId: `B-${runId}`,
      },
      data: {
        runId: runId,
      },
    };

    const JobC = {
      name: 'C',
      queueName: QueueC
      data: {
        runId: runId,
      },
      opts: {
        jobId: `C-${runId}`,
      },
      // Needs A and B
      children: [JobA, JobB],
    };

    const JobD = {
      name: 'D',
      queueName: QueueD,
      data: {
        runId: runId
      },
      opts: {
        jobId: `D-${job.id}`,
      },
      // Needs A and B completed
      children: [JobA, JobB],
    };
    const FlowC = await flowProducer.add(JobC);
    const FlowD = await flowProducer.add(JobD);

```
JobA and JobB are manually given ids since they should not be repeated for multiple dependent parents

In this situation, since both JobC and JobD share two jobs that have the same id as children, only the first added one will be triggered. 
 Is JobD not being triggered an intentional behavior? Or is this a bug?

**Comment by roggervalf** [comment]
hi @bilalshaikh42, thanks for submitting this issue, so currently one job only can have 1 parent

**Comment by ximus** [comment]
multiple parents would be great!

I have an app flow where I need to generate a PDF to s3, then notify the client via an SMS and an email.

All those steps should ideally remain loosely couple pieces of code, not reference each other directly.

Ideal I would have:

```
+--> Generate PDF (child job)
      +--> SMS (parent job)
      +--> Email PDF (parent job)
```

I guess I can link them in series with the current abilities of bullmq:

```
+--> Generate PDF (child job)
      +--> SMS (child and parent job)
            +--> Email PDF (parent job)
```

The downside here is
a) some lost efficiency: the latter two jobs could run concurrently
b) reduced resiliency: if sms job fails, then email job doesn't run, which is too bad

or for now I'll couple the email and SMS jobs into one job, losing some granularity and code separation.
or maybe there's a solution using events? listen for PDF complete and queue the two latter jobs then.

Thanks for BullMQ, a solid contribution to the node world

**Comment by manast** [comment]
@ShadowStrikerJ Unfortunatelly flows do not support multiple parents, and I do not know if and/or when they will be supported.

**Comment by imperfect-circuits** [comment]
The alternative (which we use) is to create the two new jobs within one parent job, then they can processed indepedently. E.g.

```
parent: jobParent
children: jobA, jobB

jobParent() {
   create jobC
   create jobD
}
```

**Comment by pbell23** [comment]
@imperfect-circuits This solution looks interesting but it adds complexity. If the jobC succeeds but not the jobD it will be necessary to make fail jobParent and then it will be necessary to add code to reexecute jobD and not jobC. Also this assumes that jobC and jobD have no parents as they are no longer part of the flow tree as soon as they are manually created by jobParent.

**Comment by matpen** [comment]
> @ShadowStrikerJ Unfortunatelly flows do not support multiple parents, and I do not know if and/or when they will be supported.

@manast could you expand on whether this limitation is due to specific technical reasons, or simply not implemented yet? Maybe someone in the community might want to pitch in!

**Comment by manast** [comment]
It increases complexity and edge cases maybe an order of magnitude, which use case is so important that will require this functionality?

**Comment by matpen** [comment]
> which use case is so important that will require this functionality?

Every workflow with multiple dependencies will need this functionality. In addition to the example in the OP, I will expand [the example from the documentation](https://docs.bullmq.io/guide/flows).

Currently the example looks like this:
```js
const flow = await flowProducer.add({
  name: 'renovate-interior',
  queueName: 'renovate',
  children: [
    { name: 'paint', data: { place: 'ceiling' }, queueName: 'steps' },
    { name: 'paint', data: { place: 'walls' }, queueName: 'steps' },
    { name: 'fix', data: { place: 'floor' }, queueName: 'steps' },
  ],
});
```
But suppose that, more realistically, painting the ceiling and the walls can only be approached after the floor is fixed. In this case, both paint jobs will depend on the floor fix, which in turn should have two parents.

The dependency graph now looks as follows:
![image](https://github.com/taskforcesh/bullmq/assets/9569236/1e070bcb-ac9d-4392-b7ec-2ce1a35c0e64)

I see no way to implement this situation, apart from:
* [topologically sorting](https://en.wikipedia.org/wiki/Topological_sorting) the graph, thereby loosing concurrency (the two paint jobs can no longer be run in parallel)
* coming up with a custom solution based on the [flow pattern](https://docs.bullmq.io/patterns/flows)

**Comment by manast** [comment]
Yeah, I am aware of the theoretical cases, I mean a real-world case where this is needed and no simple workaround can be found. For instance, in your example above you could just have one queue for fixing the floor, and in the worker for said queue, you can add the rest of the flow as the last step before returning.

**Comment by matpen** [comment]
@manast The case I described was a simple one just as a basis for discussion. Of course, simple workarounds can be found for this and other cases: the workaround you suggest is basically based on the [flow pattern](https://docs.bullmq.io/patterns/flows), as mentioned above.

It does have, however, the problem that it does not scale well: with a more involved dependency graph, where there are multiple nodes at every level, one must resort to topological sorting. And this will prevent the flow from maximizing concurrency. To visualize this, please consider the following image:
![image](https://github.com/taskforcesh/bullmq/assets/9569236/966a49b2-5787-4f9f-8916-853599f37c0c)

By using the flow pattern here, it is evident that we are loosing efficiency: `paint_stairs` could run as soon as `fix_stairs` is finished, but has to wait for `fix_floor` so that it can be run together with the other two paint tasks.

Please understand that here I am just trying to make the point for a genuine use case, not a "theoretical scenario". Whether such a case emerges often enough to justify the implementation effort, I would prefer to leave to the judgement of the community. I just thought to mention this, because I found many requests for this feature:
* https://github.com/taskforcesh/bullmq/issues/1021#issue-1114405232
* https://github.com/taskforcesh/bullmq/issues/1163
* https://github.com/taskforcesh/bullmq/issues/2087

That said, thank you for working on a great library!

**Comment by zaquas77** [comment]
Hi @manast and @matpen I have same problems.

In my situation I've a tools where the users define jobs and theirs depends. I like to permormance in parallels job.

Thanks again for the great work!

**Comment by zaquas77** [comment]
Hi @manast sorry for my insistence, but have you got some idea about this request? Could I help you in some way? 

thanks again.

**Comment by x-etienne** [comment]
+1

**Comment by chingiz19** [comment]
+1

**Comment by lucard17** [comment]
+1

**Comment by roggervalf** [comment]
hi guys, we are aware of this topic but we don't have an expected time to revisit it. Still this is a good issue to discuss your expectations

**Comment by jacob-israel-turner** [comment]
@manast I have a real-world use-case for you.

I have 3 jobs I need to run against a document, and 1 job that will parse and cache the document contents. The document parsing is a really expensive step, so I will not be doing it in all 3 jobs - it will be done as a pre-processing step.

The 3 jobs require the contents of the document in order to run. I need to create a flow like this:

Job A parses the document and caches it.
Job B reads the document contents and pulls out information X (only runs after Job A completes).
Job C reads the document contents and pulls out information Y (only runs after Job A completes).
Job D reads the document contents and pulls out information Z (only runs after Job A completes).

To be clear, I have not installed BullMQ in this project yet. I've used it in projects in the past (when it was still called Bull), so it was top of mind when I needed a solution for this project. However, the inflexibility of flows may be a showstopper for usage in this project. I anticipate having many more flows like this in the future, and I need a tool that will support my real-world use-cases.

**Comment by jacob-israel-turner** [comment]
> Yeah, I am aware of the theoretical cases, I mean a real-world case where this is needed and no simple workaround can be found. For instance, in your example above you could just have one queue for fixing the floor, and in the worker for said queue you can add the rest of the flow as the last step before returning.

This is a possible workaround, but it's not very reasonable. When I'm creating the complex flow at the start, I have all of the data (including sensitive credentials) that are required for the entire flow. If I have to break the flow into multiple flows, then I have to shim this data through every flow, muddying the business logic and making the flows brittle.

As it stands, the flow I'm planning would need to be broken up into at least 3 flows, unless the child-parent relationship is made more flexible.

**Comment by jacob-israel-turner** [comment]
I decided to implement this using temporal.io instead. The inflexibility of creating the workflow graph was the primary deciding factor.

**Comment by faileon** [comment]
Are there any plans for this feature? I have a user defined DAG which represents jobs that need to run. I split it into multiple flows based on my leaf nodes, but sharing parents by using job with same ID does not work :(

**Comment by manast** [comment]
@faileon there are plans, but we do not have an ETA yet.

**Comment by DGrudzynskyi** [comment]
Adding a bit regarding the demand for this feature:

In our current project we use bull (old one) with the custom shop mechanics for sequencial jobs execution.

We wanted to migrate to bullmq.pro to leverage from built-in flows and throw away the custom code nobody understands (excepts author). We also wanted to use group concurrency limits.

However, absence of the "single-child" -> "multiple-parents" and sebsequent "multiple-children" -> "multiple-parents" makes the transition unlikely because it still will not let us throw away the cheerwheel for flows.

it's not like we have the cases which are absolutely "cannot" be solved within "parent->children" but rather application is complex enough making re-arrange all business logic this way barely possible.


Hope this feature makes it path to bullmq.

**Comment by chkdmin** [comment]
Until support for multiple parents is properly implemented, it might be safer to throw an error when multiple parents are set.
